## 대출 데이터 불러오기

In [ ]:
import pandas as pd
result = pd.read_csv("C:/Users/123/Desktop/data/loan_result.csv")
result.head(3)

## 대출 결측치 개수 확인

In [2]:
result.isna().sum()

application_id                 0
loanapply_insert_time          0
bank_id                        0
product_id                     0
loan_limit                  7495
loan_rate                   7495
is_applied               3257239
dtype: int64

### 대출 칼럼 확인

In [3]:
result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13527363 entries, 0 to 13527362
Data columns (total 7 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   application_id         int64  
 1   loanapply_insert_time  object 
 2   bank_id                int64  
 3   product_id             int64  
 4   loan_limit             float64
 5   loan_rate              float64
 6   is_applied             float64
dtypes: float64(3), int64(3), object(1)
memory usage: 722.4+ MB


### 대출적용 시간 datetime으로 변환

In [4]:
# loanapply_insert_time (object -> datetime 변환)
import pandas as pd
result['loanapply_insert_time'] = pd.to_datetime(result['loanapply_insert_time'])
result.head(10)

application_id loanapply_insert_time  bank_id  product_id  loan_limit  \
0         1748340   2022-06-07 13:05:41        7         191  42000000.0   
1         1748340   2022-06-07 13:05:41       25         169  24000000.0   
2         1748340   2022-06-07 13:05:41        2           7  24000000.0   
3         1748340   2022-06-07 13:05:41        4         268  29000000.0   
4         1748340   2022-06-07 13:05:41       11         118   5000000.0   
5         1748340   2022-06-07 13:05:40       35         168  21000000.0   
6         1748340   2022-06-07 13:05:40       44           8   3000000.0   
7         1748340   2022-06-07 13:05:41       28         217  10000000.0   
8         1748340   2022-06-07 13:05:46       47         181  31000000.0   
9         1748340   2022-06-07 13:05:51       11         170  50000000.0   

   loan_rate  is_applied  
0       13.6         NaN  
1       17.9         NaN  
2       18.5         NaN  
3       10.8         NaN  
4       16.4         NaN  
5       15.2         NaN  
6       14.8         NaN  
7       18.0         NaN  
8       13.0         NaN  
9       13.3         NaN

In [5]:
result['loanapply_insert_time'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 13527363 entries, 0 to 13527362
Series name: loanapply_insert_time
Non-Null Count     Dtype         
--------------     -----         
13527363 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 103.2 MB


## application_id 와 loanapply_insert_time 으로 정렬

In [6]:
result_sorted = result.sort_values(by=['application_id','loanapply_insert_time'],ascending=True)
result_sorted.head(10)

application_id loanapply_insert_time  bank_id  product_id  \
10972632               1   2022-05-06 11:05:42       59         150   
10972629               1   2022-05-06 11:05:43       38         223   
10972631               1   2022-05-06 11:05:43       50         142   
10972630               1   2022-05-06 11:05:44       37         206   
10972628               1   2022-05-06 11:05:56       38          16   
10972627               1   2022-05-06 11:05:57       62         200   
7334795                2   2022-04-13 10:58:41       62         200   
8356349                4   2022-06-20 10:24:38        7         220   
8356360                4   2022-06-20 10:24:38        7         191   
11912958               6   2022-04-26 02:02:44       59         150   

          loan_limit  loan_rate  is_applied  
10972632  19000000.0       17.7         0.0  
10972629  19000000.0       17.1         0.0  
10972631  18000000.0       19.9         0.0  
10972630  32000000.0       18.9         0.0  
10972628  19000000.0       17.1         0.0  
10972627   8000000.0       14.8         0.0  
7334795   15000000.0        9.3         0.0  
8356349   10000000.0       11.4         NaN  
8356360   26000000.0       11.4         NaN  
11912958  19000000.0       14.0         0.0

### 한도조회 일시 시간 나누기 (24시간을 6시간으로 나누기 0시부터 6시 , 6시부터 12시, 12시부터 18시, 18시부터 24시)

In [7]:
result_insert_time  = result.set_index('loanapply_insert_time')

result_insert_time_range_0to6 = result_insert_time[(pd.DatetimeIndex(result_insert_time.index).hour >=0) & (pd.DatetimeIndex(result_insert_time.index).hour <6)]
print(len(result_insert_time_range_0to6))
result_insert_time_range_6to12 = result_insert_time[(pd.DatetimeIndex(result_insert_time.index).hour >=6) & (pd.DatetimeIndex(result_insert_time.index).hour <12)]
print(len(result_insert_time_range_6to12))
result_insert_time_range_12to18 = result_insert_time[(pd.DatetimeIndex(result_insert_time.index).hour >=12) & (pd.DatetimeIndex(result_insert_time.index).hour <18)]
print(len(result_insert_time_range_12to18))
result_insert_time_range_18to24 = result_insert_time[(pd.DatetimeIndex(result_insert_time.index).hour >=18) & (pd.DatetimeIndex(result_insert_time.index).hour <24)]
print(len(result_insert_time_range_18to24))

930032
4178599
5258354
3160378


In [8]:
data1=result_insert_time_range_0to6.iloc[:,[0,1,2,3,4]]
data1=data1.dropna(axis=0)

data2=result_insert_time_range_6to12.iloc[:,[0,1,2,3,4]]
data2=data2.dropna(axis=0)
data2.isna().sum()

data3=result_insert_time_range_12to18.iloc[:,[0,1,2,3,4]]
data3=data3.dropna(axis=0)
data3.isna().sum()

data4=result_insert_time_range_18to24.iloc[:,[0,1,2,3,4]]
data4=data4.dropna(axis=0)
data4.isna().sum()

application_id    0
bank_id           0
product_id        0
loan_limit        0
loan_rate         0
dtype: int64

### 각 시간대별로 나눈 데이터 그룹 부여

In [9]:
data1['var']=1
data2['var']=2
data3['var']=3
data4['var']=4

## 각 집단별로 대출한도 평균 차이 검정(tukey)

In [10]:
import scipy.stats as stats
import numpy as np
F_statistic, pVal = stats.f_oneway(data1['loan_limit'],data2['loan_limit'],data3['loan_limit'],data4['loan_limit'])
print('F={0:.1f}, p={1:.5f}'.format(F_statistic,pVal))

F=102.6, p=0.00000


#### F=102.6 이고 p=0.00000 이므로 집단별로 평균 차이가 있다고 판단 가능

### 2번 그룹과 3번 그룹의 차이가 없는 것 제외하고 나머지 그룹 간에는 차이가 있다고 판단.

In [11]:
df_new = pd.concat([data1,data2,data3,data4])
df_new.head(10)

from statsmodels.stats.multicomp import pairwise_tukeyhsd

posthoc = pairwise_tukeyhsd(df_new['loan_limit'], df_new['var'], alpha=0.05)
print(posthoc)

       Multiple Comparison of Means - Tukey HSD, FWER=0.05        
group1 group2   meandiff   p-adj     lower        upper     reject
------------------------------------------------------------------
     1      2  493300.4283    0.0   280551.364  706049.4925   True
     1      3  594560.6868    0.0  385823.5787  803297.7949   True
     1      4 1240418.9549    0.0 1021525.0421 1459312.8677   True
     2      3  101260.2586 0.1409  -20367.8685  222888.3857  False
     2      4  747118.5266    0.0   608782.826  885454.2273   True
     3      4   645858.268    0.0  513775.7861    777940.75   True
------------------------------------------------------------------
